File: UI/Models/SearchModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;

using Microsoft.FSharp.Control;
using Microsoft.FSharp.Core;

using Algorithm;
using System.Windows.Input;

namespace UI.Models {

    public class SearchModel : DependencyObject {

        public static readonly DependencyProperty LogNameProperty =
            DependencyProperty.Register("LogName", typeof(string), typeof(SearchModel));
        public string LogName {
            get { return (string)GetValue(LogNameProperty); }
            set { SetValue(LogNameProperty, value); }
        }

        public static readonly DependencyProperty TargetProperty =
            DependencyProperty.Register("Target", typeof(DateTime), typeof(SearchModel));
        public DateTime Target {
            get { return (DateTime)GetValue(TargetProperty); }
            set { SetValue(TargetProperty, value); }
        }

        public static readonly DependencyProperty TargetResultProperty =
            DependencyProperty.Register("TargetResult", typeof(DateTime?), typeof(SearchModel));
        public DateTime? TargetResult {
            get { return (DateTime?)GetValue(TargetResultProperty); }
            set { SetValue(TargetResultProperty, value); }
        }

        public static readonly DependencyProperty LowerBoundProperty =
            DependencyProperty.Register("LowerBound", typeof(long), typeof(SearchModel));
        public long LowerBound {
            get { return (long)GetValue(LowerBoundProperty); }
            set { SetValue(LowerBoundProperty, value); }
        }

        public static readonly DependencyProperty UpperBoundProperty =
            DependencyProperty.Register("UpperBound", typeof(long), typeof(SearchModel));
        public long UpperBound {
            get { return (long)GetValue(UpperBoundProperty); }
            set { SetValue(UpperBoundProperty, value); }
        }

        public static readonly DependencyProperty MinimumProperty =
            DependencyProperty.Register("Minimum", typeof(long), typeof(SearchModel), new UIPropertyMetadata(0L));
        public long Minimum {
            get { return (long)GetValue(MinimumProperty); }
            set { SetValue(MinimumProperty, value); }
        }

        public static readonly DependencyProperty MaximumProperty =
            DependencyProperty.Register("Maximum", typeof(long), typeof(SearchModel), new UIPropertyMetadata(100L));
        public long Maximum {
            get { return (long)GetValue(MaximumProperty); }
            set { SetValue(MaximumProperty, value); }
        }

        public static readonly DependencyProperty RangeProperty =
            DependencyProperty.Register("Range", typeof(long), typeof(SearchModel));
        public long Range {
            get { return (long)GetValue(RangeProperty); }
            set { SetValue(RangeProperty, value); }
        }

        public static readonly DependencyProperty QueryPositionProperty =
            DependencyProperty.Register("QueryPosition", typeof(long), typeof(SearchModel), new UIPropertyMetadata(0L));
        public long QueryPosition {
            get { return (long)GetValue(QueryPositionProperty); }
            set { SetValue(QueryPositionProperty, value); }
        }

        public static readonly DependencyProperty CurrentPositionProperty =
            DependencyProperty.Register("CurrentPosition", typeof(long), typeof(SearchModel), new UIPropertyMetadata(0L));
        public long CurrentPosition {
            get { return (long)GetValue(CurrentPositionProperty); }
            set { SetValue(CurrentPositionProperty, value); }
        }

        public static readonly DependencyProperty StatusProperty =
            DependencyProperty.Register("Status", typeof(string), typeof(SearchModel));
        public string Status {
            get { return (string)GetValue(StatusProperty); }
            set { SetValue(StatusProperty, value); }
        }

        public static readonly DependencyProperty IsExecutingProperty =
            DependencyProperty.Register("IsExecuting", typeof(bool), typeof(SearchModel));
        public bool IsExecuting {
            get { return (bool)GetValue(IsExecutingProperty); }
            set { SetValue(IsExecutingProperty, value); }
        }

        public static readonly DependencyProperty IsCancellingProperty =
            DependencyProperty.Register("IsCancelling", typeof(bool), typeof(SearchModel));
        public bool IsCancelling {
            get { return (bool)GetValue(IsCancellingProperty); }
            set { SetValue(IsCancellingProperty, value); }
        }

        public bool CanExecute { get { return !IsExecuting && Status != "Found"; } }

        public bool CanCancel { get { return IsExecuting && !IsCancelling; } }

        private CancellationTokenSource TokenSource { get; set; }

        private SearchState State { get; set; }

        private BinaryLogSearch Search { get; set; }

        public SearchModel(BinaryLogSearch search, DateTime target) {
            LogName = search.Name;
            Search = search;
            Target = target;
            State = new SearchState(Target, Algorithm.Status.Idle, FSharpOption<Position>.None);
        }

        public void Cancel() {
            if (IsExecuting && TokenSource != null) {
                IsCancelling = true;
                TokenSource.Cancel(false);
            }
        }

        public void Execute() {
            if (!IsExecuting) {
                IsExecuting = true;
                TokenSource = new CancellationTokenSource();
                
                var worker = new BackgroundWorker();
                worker.WorkerReportsProgress = true;
                worker.WorkerSupportsCancellation = true;
                worker.DoWork += (sender, e) => {
                    Search.Execute(TokenSource.Token, State, (state) => {
                        worker.ReportProgress(0, state);
                    });
                };
                worker.ProgressChanged += (sender, e) => {
                    State = (SearchState)e.UserState;
                    
                    if (FSharpOption<Position>.get_IsSome(State.Position)) {
                        var position = State.Position.Value;
                        if (Maximum != position.Maximum) {
                            Maximum = position.Maximum;
                        }
                        if (Minimum != position.Minimum) {
                            Minimum = position.Minimum;
                        }
                        if (LowerBound != position.LowerBound) {
                            LowerBound = position.LowerBound;
                        }
                        if (UpperBound != position.UpperBound) {
                            UpperBound = position.UpperBound;
                        }
                        if (QueryPosition != position.QueryAt) {
                            QueryPosition = position.QueryAt;
                        }
                        if (CurrentPosition != position.Current) {
                            CurrentPosition = position.Current;
                        }
                        var range = position.UpperBound - position.LowerBound;
                        if (Range != range) {
                            Range = range;
                        }
                    }

                    if (State.Status.IsIdle) {
                        Status = "Idle";
                    } else if (State.Status.IsScan) {
                        Status = "Scan";
                    } else if (State.Status.IsSeek) {
                        Status = "Seek";
                    } else if (State.Status.IsCancelled) {
                        Status = "Cancelled";
                    } else if (State.Status.IsNotFound) {
                        Status = "Not Found";
                    } else if (State.Status.IsFound) {
                        Status = "Found";
                        var found = (Algorithm.Status.Found)State.Status;
                        TargetResult = found.Item1;
                        UpperBound = found.Item2;
                        LowerBound = found.Item2;
                        QueryPosition = found.Item2;
                        CurrentPosition = found.Item2;
                        Range = 0;
                    }
                };
                worker.RunWorkerCompleted += (sender, e) => {
                    IsCancelling = false;
                    IsExecuting = false;
                    CommandManager.InvalidateRequerySuggested();
                };
                worker.RunWorkerAsync();
            }
        }
    }
}

File: UI/Models/SearchModel.fs
namespace UI.Models

open System
open System.ComponentModel
open System.Threading
open System.Windows
open System.Windows.Input

open Algorithm

type SearchModel(search : BinaryLogSearch, target : DateTime) =
    inherit DependencyObject()

    let cancel : Ref<unit->unit> = ref id
    let state : Ref<SearchState> = ref { SearchState.empty with Target=target }
    
    static let targetResult =
        DependencyProperty.Register("TargetResult", typeof<Nullable<DateTime>>, typeof<SearchModel>)
    static let lowerBound =
        DependencyProperty.Register("LowerBound", typeof<int64>, typeof<SearchModel>)
    static let upperBound =
        DependencyProperty.Register("UpperBound", typeof<int64>, typeof<SearchModel>)
    static let minimum =
        DependencyProperty.Register("Minimum", typeof<int64>, typeof<SearchModel>)
    static let maximum =
        DependencyProperty.Register("Maximum", typeof<int64>, typeof<SearchModel>)
    static let range =
        DependencyProperty.Register("Range", typeof<int64>, typeof<SearchModel>)
    static let total =
        DependencyProperty.Register("Total", typeof<int64>, typeof<SearchModel>)
    static let queryPosition =
        DependencyProperty.Register("QueryPosition", typeof<int64>, typeof<SearchModel>)
    static let currentPosition =
        DependencyProperty.Register("CurrentPosition", typeof<int64>, typeof<SearchModel>)
    static let status =
        DependencyProperty.Register("Status", typeof<string>, typeof<SearchModel>)
    static let isExecuting =
        DependencyProperty.Register("IsExecuting", typeof<bool>, typeof<SearchModel>)
    static let isCancelling =
        DependencyProperty.Register("IsCancelling", typeof<bool>, typeof<SearchModel>)

    member public x.LogName = search.Name

    member public x.Target = target

    member public x.TargetResult
        with get() = x.GetValue(targetResult) :?> Nullable<DateTime>
        and set(value:Nullable<DateTime>) = x.SetValue(targetResult, value)

    member public x.LowerBound
        with get() = x.GetValue(lowerBound) :?> int64
        and set(value:int64) = x.SetValue(lowerBound, value)

    member public x.UpperBound
        with get() = x.GetValue(upperBound) :?> int64
        and set(value:int64) = x.SetValue(upperBound, value)

    member public x.Minimum
        with get() = x.GetValue(minimum) :?> int64
        and set(value:int64) = x.SetValue(minimum, value)

    member public x.Maximum
        with get() = x.GetValue(maximum) :?> int64
        and set(value:int64) = x.SetValue(maximum, value)

    member public x.Range
        with get() = x.GetValue(range) :?> int64
        and set(value:int64) = x.SetValue(range, value)

    member public x.Total
        with get() = x.GetValue(total) :?> int64
        and set(value:int64) = x.SetValue(total, value)

    member public x.QueryPosition
        with get() = x.GetValue(queryPosition) :?> int64
        and set(value:int64) = x.SetValue(queryPosition, value)

    member public x.CurrentPosition
        with get() = x.GetValue(currentPosition) :?> int64
        and set(value:int64) = x.SetValue(currentPosition, value)

    member public x.Status
        with get() = x.GetValue(status) :?> string
        and set(value:string) = x.SetValue(status, value)

    member public x.IsExecuting
        with get() = x.GetValue(isExecuting) :?> bool
        and set(value:bool) = x.SetValue(isExecuting, value)

    member public x.IsCancelling
        with get() = x.GetValue(isCancelling) :?> bool
        and set(value:bool) = x.SetValue(isCancelling, value)

    member public x.IsFound =
        x.Status = "Found"

    member public x.CanExecute =
        not x.IsExecuting && not x.IsFound

    member public x.CanCancel =
        x.IsExecuting && not x.IsCancelling

    member public x.Cancel() = (!cancel)()

    member public x.Execute() =
        if x.CanExecute then
            x.IsExecuting <- true

            let tokenSource = new CancellationTokenSource()

            cancel := fun () ->
                if x.CanCancel then
                    x.IsCancelling <- true
                    tokenSource.Cancel(false)

            let worker = new BackgroundWorker()
            worker.WorkerReportsProgress <- true
            worker.WorkerSupportsCancellation <- true
            worker.DoWork.Add(fun e ->
                search.Execute(tokenSource.Token, !state, fun state ->
                    worker.ReportProgress(0, state))
                |> ignore)
            worker.ProgressChanged.Add(fun e ->
                state := e.UserState :?> SearchState
                let state = !state

                match state.Position with
                | None -> ()
                | Some position ->
                    if x.Maximum <> position.Maximum then
                        x.Maximum <- position.Maximum

                    if x.Minimum <> position.Minimum then
                        x.Minimum <- position.Minimum

                    if x.LowerBound <> position.LowerBound then
                        x.LowerBound <- position.LowerBound

                    if x.UpperBound <> position.UpperBound then
                        x.UpperBound <- position.UpperBound

                    if x.QueryPosition <> position.QueryAt then
                        x.QueryPosition <- position.QueryAt

                    if x.CurrentPosition <> position.Current then
                        x.CurrentPosition <- position.Current

                    let range = position.UpperBound - position.LowerBound
                    if x.Range <> range then
                        x.Range <- range

                    let total = position.Maximum - position.Minimum
                    if x.Total <> total then
                        x.Total <- total

                let status = 
                    match state.Status with
                    | Idle ->
                        "Idle"
                    | Scan ->
                        "Scan"
                    | Seek ->
                        "Seek"
                    | Cancelled ->
                        "Cancelled"
                    | NotFound ->
                        "NotFound"
                    | Found (at, index) ->
                        x.TargetResult <- Nullable(at)
                        x.UpperBound <- index
                        x.LowerBound <- index
                        x.QueryPosition <- index
                        x.CurrentPosition <- index
                        x.Range <- 0L
                        "Found"
                if x.Status <> status then
                    x.Status <- status)
            worker.RunWorkerCompleted.Add(fun e ->
                x.IsCancelling <- false
                x.IsExecuting <- false
                if not(isNull e.Error) then
                    x.Status <-
                        if e.Error.Message.Contains "Invalid JSON" then "Invalid JSON"
                        else sprintf "Error: %s" e.Error.Message)
            worker.RunWorkerAsync()
